components/hts221/hts221.h

```diff
@@ -0,0 +1,201 @@
+#define HTS221_REG_TEMP_OUT_H 0x2B   // R   Temperature output register (MSB)
+#define HTS221_REG_CALIB_START 0x30  // R/W Calibration start register
+
+#define HTS221_AV_CONF_DEFAULT \
+  0x1B  // DEFAULT AV_CONF status register value according to datasheet
+
+#define HTS221_CTRL_REG1 \
+  0x85  // Set CTRL_REG1 to power on mode, enable block data update, set output
+        // data rate to 1Hz
+#define HTS221_CTRL_REG1_POWERON \
+  0x80  // CTRL_REG1 status register value to power on HTS221
+#define HTS221_CTRL_REG1_POWERDOWN \
+  0x00  // CTRL_REG1 status register value to power on HTS221
+
+#define HTS221_CTRL_REG2 \
+  0x00  // Default CTRL_REG2 status register
+        // value according to datasheet
+#define HTS221_CTRL_REG2_HEATER_ON \
+  0x02  // CTRL_REG2 status register value to enable heater
+#define HTS221_CTRL_REG2_HEATER_OFF \
+  0x00  // CTRL_REG2 status register value to
+        // disable heater
+
+#define HTS221_CTRL_REG3 \
+  0x00  // Default CTRL_REG3 status register value according to datasheet
+
+/**
+ * @brief Struct for storing calibration parameters.
+ *
+ * Unique for every HTS221 Sensor.
+ */
+typedef struct {
+  uint8_t H0_rH;
+  uint8_t H1_rH;
+  uint16_t T0_degC;
+  uint16_t T1_degC;
+  int16_t H0_T0_OUT;
+  int16_t H1_T0_OUT;
+  int16_t T0_OUT;
+  int16_t T1_OUT;
+} calibration_param_t;
+
+/**
+ * @brief Initialize the HTS221 Device descriptor
+ *
+ * @param[out] dev Device descriptor
+ * @param[in] addr I2C address
+ * @param[in] port I2C port number
+ * @param[in] sda_gpio GPIO pin number for SDA
+ * @param[in] scl_gpio GPIO pin number for SCL
+ * @return `ESP_OK` on success
+ */
+esp_err_t hts221_init_desc(i2c_dev_t *dev, uint8_t addr, i2c_port_t port,
+                           gpio_num_t sda_gpio, gpio_num_t scl_gpio);
+
+/**
+ * @brief Free device descriptor
+ *
+ * @param[in] dev Device descriptor
+ * @return `ESP_OK` on success
+ */
+esp_err_t hts221_free_desc(i2c_dev_t *dev);
+
+/**
+ * @brief Read the HTS221 WHO_AM_I register (0xBC)
+ *
+ * @param[in] dev Device descriptor
+ * @param[out] who_am_i Device identifier, should be (OxBC)
+ * @return `ESP_OK` on success
+ */
+esp_err_t hts221_who_am_i(i2c_dev_t *dev, uint8_t *who_am_i);
+
+/**
+ * @brief Setup device parameters
+ *
+ * Set the AV_CONF register to the default resolution mode (0x1B) according to
+ * the datasheet.
+ * Set CTRL_REG1 to power on mode, enable block data update, set output
+ * data rate to 1Hz.
+ * Set CTRL_REG2 to default value (0x00), BOOT bit normal, heater off, one-shot
+ * set to 0.
+ *
+ * @param[in] dev Device descriptor
+ * @return `ESP_OK` on success
+ */
+esp_err_t hts221_setup(i2c_dev_t *dev);
+
+/**
+ * @brief Read the HTS221 humidity and temperature resolution mode register
+ *
+ * @param[in] dev Device descriptor
+ * @param[out] av_conf humidity and temperature resolution mode register value,
+ * default (0x1B)
+ * @return `ESP_OK` on success
+ */
+esp_err_t hts221_read_av_conf(i2c_dev_t *dev, uint8_t *av_conf);
```

Comment:
What is `av_conf`? Smells like Arduino. I strongly dislike this approach. Why then do you need a driver at all?
1. Create two enumerations: `hts221_temperature_avg_t` and `hts221_humidity_avg_t`
2. Declare this function as follows: 
```C
esp_err_t hts221_read_av_conf(i2c_dev_t *dev, hts221_temperature_avg_t *t_avg, hts221_humidity_avg_t *rh_avg);
```
The same applies to other similar functions.